ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPool.java

```diff
@@ -121,13 +129,13 @@ private void initLogForReplay() {
         .log();
   }
 
-  public void saveToDisk() {
+  private synchronized void saveToDisk(final PendingTransactions pendingTransactionsToSave) {
```

Comment:
Why pick two arbitrary methods to place the `synchronized` keyword on.  What fields are they looking to prevent unsynchronized access on?  Do all accesses to those fields get locked?

Consider using `com.google.errorprone.annotations.concurrent.GuardedBy` annotations.

Reply:
I just want that these 2 methods are executed serially, in case of different threads enabling/disabling the pool in rapid succession (that should never happen now).

Reply:
That's my point, we are treating synchronization lightly.  

If they shouldn't be executed concurrently then you should be able to identify the fields that would conflict.  Then we can make sure that no other public method will allow the same conflicts that executing these two methods in parallel would cause.

Consider using `import java.util.concurrent.locks.ReentrantLock` if your goal is to prevent writing and reading at the same time.  This will limit the impact of the lock (not locking on the entire TransactionPool object, which may have unforseen impacts).  It also has timeout-enabled lock acquisitions, which should be considered to ensure shutdown doesn't fully freeze in extreme circumstances, or take an extended time such as if multiple shutdown requests have been initiated.

Reply:
Ok, I see your point, will change to avoid locking the entire class, and will change the write to check and kill if an already running write, same for the read, so there is no unbounded queueing on shutdown and it can finish in few seconds as for my tests with >100k txs. At that point there should be no need to put a timeout but in case its place should be in the `Runner::stop`.

Reply:
Please check the updated implementation

Reply:
Use the timeout variant, with something large like a minute or ten minutes.  A bug in the loading code could prevent orderly shutdown of the entire service. And this is the sort of bug that will sneak in as an unrelated change to some other feature, not that it exists today.

It's better to have a shutdown missing pending transactions than a service that refuses to shutdown.  A non-timeout lock request creates that risk.

Reply:
Managing the timeout on shutdown in `Runner::stop` where all timeout on services are managed then.